js/sagas/onboarding.js

```diff
@@ -0,0 +1,83 @@
+/**
+ * A saga that manages the Onboarding.
+ *
+ * @flow
+ */
+
+import { type Saga } from 'redux-saga'
+import { takeLatest, select, take, call, put } from 'redux-saga/effects'
+import { NavigationActions } from 'react-navigation'
+
+import { type GlobalState } from '../reducers/types'
+import {
+  SESSION_INITIALIZE_SUCCESS,
+  TOS_ACCEPT_REQUEST,
+  TOS_ACCEPT_SUCCESS,
+  PIN_CREATED_REQUEST
+} from '../store/actions/constants'
+import ROUTES from '../navigation/routes'
+
+// The Pin step of the Onboarding
+function* pinStep(): Saga<void> {
+  // From the state check ir the user already created a Pin
+  const isPinCreated: boolean = yield select(
+    (state: GlobalState): boolean => state.onboarding.isPinCreated
+  )
+
+  if (!isPinCreated) {
+    // If the Pin has not been created yet we must show the user the PinScreen
+    const navigateAction = NavigationActions.reset({
+      index: 0,
+      actions: [
+        NavigationActions.navigate({ routeName: ROUTES.ONBOARDING_PIN })
+      ],
+      key: ROUTES.ONBOARDING
+    })
+    yield put(navigateAction)
+
+    // Here we wait the user to create a Pin
+    yield take(PIN_CREATED_REQUEST)
+  }
+}
+
+// The ToS step of the Onboarding
+function* tosStep(): Saga<void> {
+  // From the state we check if the user already accepted
+  const isTosAccepted: boolean = yield select(
+    (state: GlobalState): boolean => state.onboarding.isTosAccepted
+  )
+
+  if (!isTosAccepted) {
+    // If the ToS is not accepted we must to show the user the TosScreen
+    const navigateAction = NavigationActions.reset({
+      index: 0,
+      actions: [
+        NavigationActions.navigate({ routeName: ROUTES.ONBOARDING_TOS })
+      ],
+      key: ROUTES.ONBOARDING
+    })
+    yield put(navigateAction)
+
+    // Here we wait the user accept the ToS
+    yield take(TOS_ACCEPT_REQUEST)
+
+    // Dispatch the action that sets isTosAccepted to true into the store
+    yield put({
+      type: TOS_ACCEPT_SUCCESS
+    })
+
+    // Go to the next step
+    yield call(pinStep)
```

Comment:
I'm not a saga expert so I'm asking here: doesn't this way of calling saga steps directly introduce tight coupling between sagas ? I mean, can the pinStep saga just run taking(TOS_ACCEPT_SUCCESS) instead of being called here ?

Reply:
I am also still experimenting with saga, but your objection is really valid. I am going to try to make each step listening for a specific action instead of getting called directly and check if it all works in that way.

Reply:
moreover, everywhere is possibile, I'd pass the data needed into the action payload so you don't have to yield select() into the state. ie `function* pinStep(isPinRequired: boolean): Saga<void>`
which makes testing easier. this works with call() as well (you can pass parameters)

Reply:
I was thinking about your suggestion of using take(TOS_ACCEPT_SUCCESS, pinStep) to start the pinStep saga. The problem is that if the user close the App after accepting the Tos when the App is reopened the TOS_ACCEPT_SUCCESS will never be dispatched. This is why i was calling the pinStep directly. The solution could be to dispatch from the tosStep a 'TOS_ALREADY_ACCEPTED' (if tos already accepted) then use take([TOS_ACCEPT_SUCCESS, TOS_ALREADY_ACCEPTED], pinStep). What do you think?

Reply:
sounds good. I favor every refactor that strives to accomplish these things:
- avoid calling yield select() to [make sagas self-consistent](https://github.com/redux-saga/redux-saga/tree/master/docs/api#notes-13)
- pass all the needed data in action payloads
- avoid calling a saga from another saga (yield call)

Reply:
another option (don't know if it's doable) is to move the logic at the caller level (root saga "orchestrator"). I guess it's doable if between application re-starts the state is saved so we know that  isTosAccepted is true without dispatching any action.